api/service/explorer/storage.go

```diff
@@ -69,78 +103,239 @@ func (storage *Storage) Init(ip, port string) {
+			utils.Logger().Error().Err(err).Str("address", string(address)).
+				Msg("[explorer] failed to update address")
+			return err
+		}
+	}
+	for address, txRecords := range stkTxs {
+		if err := storage.UpdateTxAddressStorage(address, txRecords, true /* isStaking */); err != nil {
+			utils.Logger().Error().Err(err).Str("address", string(address)).
+				Msg("[explorer] failed to update address")
+			return err
 		}
 	}
+	storage.dirtyBNs[bn] = struct{}{}
+	return nil
+}
 
-	address.ID = addr
+// UpdateTxAddressStorage updates specific addr tx Address.
+func (storage *Storage) UpdateTxAddressStorage(addr oneAddress, txRecords TxRecords, isStaking bool) error {
+	address, err := storage.getAddressInfo(addr)
+	if err != nil {
+		return err
+	}
+
+	address.ID = string(addr)
 	if isStaking {
 		address.StakingTXs = append(address.StakingTXs, txRecords...)
 	} else {
 		address.TXs = append(address.TXs, txRecords...)
 	}
-	encoded, err := rlp.EncodeToBytes(address)
-	if err == nil {
-		storage.GetDB().Put([]byte(key), encoded, nil)
-		storage.addrIndexCache.Add(key, address)
-	} else {
-		utils.Logger().Error().
-			Bool("isStaking", isStaking).Err(err).Msg("cannot encode address")
+	storage.dirty[addr] = address
+	return nil
+}
+
+func (storage *Storage) flushLocked() error {
+	storage.lock.RLock()
+	batch := new(leveldb.Batch)
+	for addr, addressInfo := range storage.dirty {
+		key := GetAddressKey(addr)
+		encoded, err := rlp.EncodeToBytes(addressInfo)
+		if err != nil {
+			utils.Logger().Error().Err(err).Msg("error when flushing explorer")
+			return err
+		}
+		batch.Put(key, encoded)
+	}
+	for bn := range storage.dirtyBNs {
+		key := GetCheckpointKey(new(big.Int).SetUint64(bn))
+		batch.Put(key, []byte{})
+	}
+	storage.lock.RUnlock()
+
+	// Hack: during db write, dirty is read-only. We can release the lock for now.
+	if err := storage.db.Write(batch, nil); err != nil {
+		return errors.Wrap(err, "failed to write explorer data")
+	}
+	// clean up
+	storage.lock.Lock()
+	defer storage.lock.Unlock()
+	for addr, addressInfo := range storage.dirty {
+		storage.clean.Add(addr, addressInfo)
+	}
+	storage.dirty = make(map[oneAddress]*Address)
+	storage.dirtyBNs = make(map[uint64]struct{})
+	return nil
+}
+
+// GetAddressInfo get the address info of the given address
+func (storage *Storage) GetAddressInfo(addr string) (*Address, error) {
+	storage.lock.RLock()
+	defer storage.lock.RUnlock()
+
+	return storage.getAddressInfo(oneAddress(addr))
+}
+
+func (storage *Storage) getAddressInfo(addr oneAddress) (*Address, error) {
+	if addrInfo, ok := storage.dirty[addr]; ok {
+		return addrInfo, nil
+	}
+	if x, ok := storage.clean.Get(addr); ok {
+		return x.(*Address), nil
+	}
+	addrInfo, err := storage.readAddressInfoFromDB(addr)
+	if err != nil {
+		if errors.Is(err, leveldb.ErrNotFound) {
+			return &Address{
+				ID: string(addr),
+			}, nil
+		}
+		return nil, err
+	}
+	storage.clean.Add(addr, addrInfo)
```

Comment:
Seems the storage.clean never gets cleared? The clean seems not like cache for data that's already committed to db?

Reply:
Clean is an LRU cache, it has it's own evict rules implemented. What do you mean by "seem not like cache"?

Reply:
I see. It is a cache. I thought it's a map like dirty.